### PR TITLE
Add failing test for simultaneous forwards- and backwards pagination

### PR DESCRIPTION
**Pull request by mxstbr**
This fails for some reason. :cry: Reference PR: #91

The test essentially boils down to fetching page 2 first, then page 3, then page 1.

### CODE REVIEW

File: src/extras/relayPagination.test.ts [hunk 1]
```diff
@@ -242,3 +242,131 @@ it('falls back to cursor field instead of PageInfo cursors', () => {
...
+    { query: Pagination, variables: { after: '2', first: 1 } },
+    pageThree
+  );
+  write(
+    store,
+    { query: Pagination, variables: { before: '1', last: 1 } },
+    pageOne
+  );
+
+  const res = query(store, { query: Pagination });
```
Comment (JoviDeCroock) [inline src/extras/relayPagination.test.ts]: You're not using any variables here (no first or last)

**Comment by mxstbr** [comment]
Great catch, added! Still failing though :cry:

**Comment by JoviDeCroock** [comment]
This is failing because the cache has no way of seeing how the pagination changed, we have no way of knowing what first: 3 is.

When we change the amount we want to see we move to another array, example: this will still result in an erroneous result but it will have all entities:

```  write(
    store,
    { query: Pagination, variables: { first: 1 } },
    pageOne
  );

  const res = query(store, { query: Pagination, variables: { first: 1 } });```

Changing these lines gives me a result where only additionalInfo is wrong

**Comment by mxstbr** [comment]
Well yeah of course `first: 1` is going to work because there is no overlap then. The first write adds `one`, the second `two` and then the actual query just fetches `one`, which is not what this test is testing.

**Comment by JoviDeCroock** [comment]
yes but first: 3 has no match, when our limit changes we can't trust the data afaik. We have to stay on the limit we're using. first implies we are going forward while last implies we are going backwards.

We should match on the the limit we are writing with

**Comment by mxstbr** [comment]
I have updated the test to pass the same arguments and limits as the writes do, and it's still broken. Does my test make sense?

**Comment by JoviDeCroock** [comment]
I personally haven't used this form of pagination yet so I can't say if this is a use case that it is meant for, the problem I'm seeing in the current iteration of our helper is that when we see last is that we will backwards navigate.

This implies that we hit our first write and then try to go backwards, this means that the next query will be (before: "1") instead of (after: "1") like we are doing now.

That's why my first comment about using "first" instead of "last" probably confused the both of us and expressed the wrong intent from my side.

**Comment by kitten** [comment]
Fixed up and merged into the mainline branch (passing now)
- https://github.com/FormidableLabs/urql-exchange-graphcache/pull/91/commits/5986de05f3f7116cfef1fae9dd4406d4562c2dba
- https://github.com/FormidableLabs/urql-exchange-graphcache/pull/91/commits/1ce52575c98194da384c0d70143c8f5b4f4f8402

**Pull request closed without merging**